### PR TITLE
Introduce partial chain recalculation

### LINKED ISSUE #1358: Add API for state recalculation
If we need to retrieve the state up to block `n`, we can do it through `BlockChain<T>.GetState(key, n.Hash)` if we have a full state of `n`. but if we don't have it, we should use `BlockChain<T>.GetState(key, n.Hash, StateCompleter.Recalculate)`.

But problem is, recalculating via `StateCompleter.Recalculate` can be time-consuming and there is no way for users to know this in advance. so it would be better if we have an explicit and retryable API for this situation.

### PR DESCRIPTION
**Pull request by greymistcube**
Closes #1358.

This addes `StateCompleterSet<T>.ComplementAll` and `StateCompleterSet<T>.ComplementLatest` in addition to existing `StateCompleterSet<T>.Recalculate`.

I strongly suspect `StateCompleterSet<T>.Recalculate` is no longer necessary but is left to remain to have minimal affect on the overall API.

Few things to note:
- `StateCompleterSet<T>.Recalculate`'s internal behavior has been changed.
  - Delegated methods now terminate early on reaching the provided hash value. Before, the argument was ignored and full complementation was forced regardless.
  - Does not skip over re-evaluating states already in the storage.
    - Does not overwrite existing states in the storage. Not sure whether it'd be better or not to.
- `StateCompleterSet<T>.ComplementAll` behaves more like the old `StateCompleterSet<T>.Recalculate`.
  - The only difference is this also has early termination on reaching the provided hash value.
- `StateCompleterSet<T>.ComplementLatest` only fills in the necessary intermediate states by finding the last stored states by iterating over blocks in reverse.

### CODE REVIEW

File: Libplanet/Blockchain/BlockChain.Recalculate.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                    break;
+                }
+                else
+                {
+                    stack.Push(p);
+                    pointer = this[p].PreviousHash;
+                }
+            }
+
+            foreach (BlockHash hash in stack)
```
Comment (limebell) [inline Libplanet/Blockchain/BlockChain.Recalculate.cs:146]: It looks like `ComplementAllBlockStates()` recalculates from older blocks to newer blocks, but this stack stored in inverse way. It is okay with that?

Comment (greymistcube) [reply Libplanet/Blockchain/BlockChain.Recalculate.cs:146]: When iterating over `Stack<T>`, iteration is done in popping order (without actually popping), i.e. how stacks would usually be consumed.

File: Libplanet.Tests/Blockchain/BlockChainTest.cs [hunk 2]
```diff
@@ -1637,14 +1673,25 @@ void BuildIndex(Guid id, Block<DumbAction> block)
...
+                trieStateStore.PruneStates(
+                    ImmutableHashSet<BlockHash>.Empty
+                        .Add(presentBlocks[0].Hash)
+                        .Add(presentBlocks[1].Hash));
             }
 
+            // Just in case
+            Assert.True(stateStore.ContainsBlockStates(presentBlocks[0].Hash));
+            Assert.True(stateStore.ContainsBlockStates(presentBlocks[1].Hash));
+            Assert.False(stateStore.ContainsBlockStates(chain.Tip.Hash));
```
[the remark this answers is not in the document]
Comment (greymistcube) [reply Libplanet.Tests/Blockchain/BlockChainTest.cs:1694]: I see your point. What would you prefer instead?

- Just remove it.
- Make a test code for a test code separately to see if resulting list fixture is what we want.
- Duplicate pre-testing to all tests making a call to this method.

[the remark this answers is not in the document]
Comment (limebell) [reply Libplanet.Tests/Blockchain/BlockChainTest.cs:1694]: I guess the first one will be enough.

[the remark this answers is not in the document]
Comment (greymistcube) [reply Libplanet.Tests/Blockchain/BlockChainTest.cs]: Done. See 17f4b816d74b0f1389e621d1793f14541607fcfc.

**Review by limebell — APPROVED** [review]

**Pull request merged**